File: subdomain-visit-count/subdomain-visit-count.py
#leetcode - Subdomain visit count - Medium

class Solution:
    def subdomainVisits(self, cpdomains: List[str]) -> List[str]:
        lookup = {}
        for domain in cpdomains:
            count_paired = domain.split()
            count = int(count_paired[0])
            whole_domain = count_paired[1]
            lookup[whole_domain] = lookup.get(whole_domain, 0) + count
            i = 0
            while i < len(whole_domain):
                if whole_domain[i] == ".":
                    sub_domain = whole_domain[i+1:]
                    lookup[sub_domain] = lookup.get(sub_domain, 0) + count
                    i += 2
                i += 1
        result = []
        for domain, count in lookup.items():
            pair = " ".join([str(count), domain])
            result.append(pair)
        return result

File: subdomain-visit-count/subdomain-visit-count.js
//leetcode - Subdomain visit count - Medium

/**
 * @param {string[]} cpdomains
 * @return {string[]}
 */

var domainToSubdomains = function(domain) {
    const result = [domain];
    let subdomain;
    for (let i = 1; i < domain.length; i++) {
        if (domain[i] === ".") {
            subdomain = domain.slice(i + 1);
            result.push(subdomain)
        }
    }
    return result
} 
var subdomainVisits = function(cpdomains) {
    const lookup = new Map();
    for (const pair of cpdomains) {
        const [count, domain] = pair.split(" ");
        for (const subdomain of domainToSubdomains(domain)) {
            if (lookup.has(subdomain)) {
                lookup.set(subdomain, lookup.get(subdomain) + Number(count))
            } else {
                lookup.set(subdomain, Number(count))
            }
        }
    }
    //const result = []
    //for (const [key, value] of lookup.entries()) {
    //    result.push(`${value} ${key}`)
    //}
    return [...lookup.entries()].map(([key, value]) => `${value} ${key}`)
    //return result;
};